modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceContextImpl.java

```diff
@@ -136,24 +140,70 @@ ExecutorService executor() {
      * @return Method.
      */
     @Nullable Method method(GridServiceMethodReflectKey key) {
-        Method mtd = mtds.get(key);
+        IgniteBiTuple<Method, HistogramMetricImpl> mtdRecord = methodRecord(key);
+
+        Method mtd = mtdRecord.get1();
 
         if (mtd == null) {
-            try {
-                mtd = svc.getClass().getMethod(key.methodName(), key.argTypes());
+            synchronized (mtdRecord) {
+                mtd = mtdRecord.get1();
 
-                mtd.setAccessible(true);
-            }
-            catch (NoSuchMethodException ignored) {
-                mtd = NULL_METHOD;
-            }
+                if (mtd == null) {
+                    try {
+                        mtd = svc.getClass().getMethod(key.methodName(), key.argTypes());
 
-            mtds.put(key, mtd);
+                        mtd.setAccessible(true);
+                    }
+                    catch (NoSuchMethodException ignored) {
+                        mtd = NULL_METHOD;
+                    }
+
+                    mtdRecord.set1(mtd);
+                }
+            }
         }
 
         return mtd == NULL_METHOD ? null : mtd;
     }
 
+    /**
+     * @param key Method key.
+     * @param histogrammInitiator Histogramm supplier if the histogramm isn't initialized yet.
+     * @return Service method performance histogramm.
+     */
+    @Nullable HistogramMetricImpl invokeHistogramm(GridServiceMethodReflectKey key,
+        Supplier<HistogramMetricImpl> histogrammInitiator) {
+        IgniteBiTuple<Method, HistogramMetricImpl> mtdRecord = methodRecord(key);
+
+        HistogramMetricImpl histogramm = mtdRecord.get2();
+
+        if (histogramm == null) {
```

Comment:
I think we can avoid double-checked locking by initiating on map#computeIfAbsent.

Reply:
I don't think so. This record might be created not by call for historgam, but by call for GridServiceMethodReflectKey. The record would exist without histogram.

Also
assert histogramm != null;
is useful.

I'll try to split the map.

Reply:
Refactored.